### PR TITLE
tests: Convert String to TextBlock

### PR DESCRIPTION
**Pull request by jukzi**
automated Cleanup

### CODE REVIEW

**Comment by mpalat** [comment]
@jukzi - We always wanted to avoid this automated cleanup since it becomes tough to figure out which test case was committed for which bug/issue. I would request to revert this commit.
@stephan-herrmann @jarthana @srikanth-sankaran - others -  your comments welcome.

**Comment by jukzi** [comment]
@mpalat sorry i did not know. Also i don't know how TextBlocks would make that task harder as it's still visible in the git history of the function. Feel free to revert - but i guess that revert commit would make the git history even more complicated.

**Comment by stephan-herrmann** [comment]
First things first: I think this commit should never have been merged, but certainly a regular revert only doubles the damage. So please @jukzi do the necessary research to find out whether we have a realistic option to rewriting git history by some kind of force push to master thus eliminating this commit.

**Comment by stephan-herrmann** [comment]
> ... Also i don't know how TextBlocks would make that task harder as it's still visible in the git history of the function ...

you could have asked anybody actively working with these tests. If you look at the git history of the major portion of this all - https://github.com/eclipse-jdt/eclipse.jdt.core/commits/master/org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression - its easy to see that despite all the changes in JDT, you basically see four names only. Asking any of the four would have given you excellent feedback on your proposal.

To spell out the damage: It's not only the question which bug report produced a given test case. For many investigations in compiler development it is essential to track down how error messages from the compiler evolved over time. So a typical query concerning a test case in question is: has the compiler output for this test changed over time? Starting from this change, the answer to each of these queries will invariably be: yes, the expected compiler output has changed, but wait, it was only @jukzi  refactoring string concatenation to TextBlocks. We will have to continuously pay this price from now on until the end of JDT. So even if the effort per instance might be acceptable, multiply by infinity :smile: to get the real price.

**Comment by iloveeclipse** [comment]
> necessary research to find out whether we have a realistic option to rewriting git history by some kind of force push to master thus eliminating this commit

I **believe** every committer has this possibility to directly force push to master.

The will be probably problem on jenkins /SDK build which will be unable to pull (as the branch history will be different).

@sravanlakkimsetti, @akurtakov: any experience with rewriting history?

@eclipse-jdt/eclipse-jdt-committers : please don’t merge anything to jdt until we have decided how to continue.

**Comment by akurtakov** [comment]
@iloveeclipse I remember you did smth like it https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/1663

**Comment by iloveeclipse** [comment]
> @iloveeclipse I remember you did smth like it https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/1663

Cool, I already forgot about it. And yes, naive attempt to force push to master was declined:
```
Repository https://github.com/eclipse-jdt/eclipse.jdt.core.git
protected branch hook declined
error: GH006: Protected branch update failed for refs/heads/master.
error: Changes must be made through a pull request. Cannot force-push to this branch
```
I will open IT ticket & try to force push.

**Comment by iloveeclipse** [comment]
Created https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/4651.

I assume all what we need once we have push right is to force push eddfb7d3c89c97db3da769539614497c8982c235 to master.
Once it is done, everyone who already pulled from master with this PR commit should hard reset local repo to commit eddfb7d3c89c97db3da769539614497c8982c235, and that should be it.

**Comment by iloveeclipse** [comment]
@eclipse-jdt/eclipse-jdt-committers : I believe we should be fine now. Please verify.

Note: if you have already pulled from master, hard reset your master branch to https://github.com/eclipse-jdt/eclipse.jdt.core/commit/eddfb7d3c89c97db3da769539614497c8982c235.

**Comment by stephan-herrmann** [comment]
BTW, this change caused regression in org.eclipse.jdt.core.tests.compiler.regression.ScannerTest.test034()

Looks like a bug in the refactoring, please follow up with JDT/UI to fix this.

**What disturbs me more:** test results from I-builds only show 5 tests from ScannerTest, out of 86 existing tests. And this has been like this for quite a while.

**Comment by iloveeclipse** [comment]
> **What disturbs me more:** test results from I-builds only show 5 tests from ScannerTest, out of 86 existing tests. And this has been like this for quite a while.

Could you please open a dedicated ticket for this?

**Comment by stephan-herrmann** [comment]
> @eclipse-jdt/eclipse-jdt-committers : I believe we should be fine now. Please verify.
> 
> Note: if you have already pulled from master, hard reset your master branch to [eddfb7d](https://github.com/eclipse-jdt/eclipse.jdt.core/commit/eddfb7d3c89c97db3da769539614497c8982c235).

Looks good. Thank you so much!

Apparently the tags I20240517-1800 and onwards still point to the commit in question (now outside of master), but I guess / hope this doesn't hurt.

**Comment by stephan-herrmann** [comment]
> > **What disturbs me more:** test results from I-builds only show 5 tests from ScannerTest, out of 86 existing tests. And this has been like this for quite a while.
> 
> Could you please open a dedicated ticket for this?

I'm still trying to figure out at which level this might be broken. Is it JDT or the SDK build?

**Comment by stephan-herrmann** [comment]
> > > **What disturbs me more:** test results from I-builds only show 5 tests from ScannerTest, out of 86 existing tests. And this has been like this for quite a while.
> > 
> > 
> > Could you please open a dedicated ticket for this?
> 
> I'm still trying to figure out at which level this might be broken. Is it JDT or the SDK build?

430 tests in https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/master/222/testReport/org.eclipse.jdt.core.tests.compiler.regression/ScannerTest/ (86 tests at 5 compliance levels).

**Comment by iloveeclipse** [comment]
> I'm still trying to figure out at which level this might be broken. Is it JDT or the SDK build?

  **If** this is broken at all, it probably broken at the test report or creation level (`org.eclipse.jdt.core.tests.junit.extension.TestCase.buildTestsList(Class, int, long)` ?), I see that there **are** tests from `ScannerTest` like `testBug531716_001_since_13` but they are marked as `org.eclipse.jdt.core.tests.compiler.regression.TestAll`.

See 
- https://download.eclipse.org/eclipse/downloads/drops4/I20240519-1800/testresults/html/org.eclipse.jdt.core.tests.compiler_ep432I-unit-cen64-gtk3-java17_linux.gtk.x86_64_17.html
- https://download.eclipse.org/eclipse/downloads/drops4/I20240519-1800/testresults/xml/org.eclipse.jdt.core.tests.compiler_ep432I-unit-cen64-gtk3-java17_linux.gtk.x86_64_17.xml

**Comment by srikanth-sankaran** [comment]
> @eclipse-jdt/eclipse-jdt-committers : I believe we should be fine now. Please verify.
> 
> Note: if you have already pulled from master, hard reset your master branch to [eddfb7d](https://github.com/eclipse-jdt/eclipse.jdt.core/commit/eddfb7d3c89c97db3da769539614497c8982c235).

And I guess if your fork's master is synched with https://github.com/eclipse-jdt/eclipse.jdt.core, you also need to force push after hard reset of master to [eddfb7d](https://github.com/eclipse-jdt/eclipse.jdt.core/commit/eddfb7d3c89c97db3da769539614497c8982c235) 

Thanks @iloveeclipse for taking care of this.

@eclipse-jdt/eclipse-jdt-committers : I would like that we adopt some long term process protocols from this. Firstly I recall reading somewhere that large scale refactoring changes are to be taken up only in M1. Unless rescinded, let us adhere to that. 

Secondly as with committer election, I would suggest a voting process for "cosmetic refactorings." - this would allow folks to weigh in on any concerns they may have before too much investment happens from the champion's side.

**Comment by jukzi** [comment]
> BTW, this change caused regression in org.eclipse.jdt.core.tests.compiler.regression.ScannerTest.test034()

Which? I do not see a test failure in  CI, i-build nor when i execute the test locally after cleanup
image

@stephan-herrmann please explain.

**Comment by jukzi** [comment]
@stephan-herrmann may be you applied the cleanup manually with an old version?
There was an error that was already fixed due to this PR: https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2315

**Comment by iloveeclipse** [comment]
> First things first: I think this commit should never have been merged

> FWIW, have sent [a mail ](https://www.eclipse.org/lists/jdt-dev/msg02377.html) to jdt-dev to make sure that people are aware of the general practice we follow in jdt.core

I believe there are at least two problems we had here.

1) Every JDT committer should have received emails about new PR's created. However, no one complained here **before PR got merged**. I think this is not OK. I personally didn't reacted on this PR as I was not aware about the "unwritten" rule that **tests** shouldn't be mass-refactored and I was assuming if the change is safe and all tests are green we are fine with that, because it's only about tests. Now I've learned that this is not the case, and here we are at the second issue.
2) https://github.com/eclipse-jdt/.github/blob/main/CONTRIBUTING.md#jdtcore-general-practices (which is updated now) **still** doesn't mention that. It would be good to add this explicit requirement there. Probably it could be generalized to something that says "no mass cleanups can be accepted if they prevent git blame from proper working". @stephan-herrmann : could you please take this action and formulate something in the https://github.com/eclipse-jdt/.github/blob/main/CONTRIBUTING.md#jdtcore-general-practices?

**Comment by jukzi** [comment]
Please also note that the CONTRIBUTING.md belongs to the whole JDT - not only jdt.core. There was explicit approval from jdt.ui lead for this PRs. 
To me it feels this discussions are a little bit overreactions. In general it is very clear to all of us that jdt.core does not like mass refactorings at all.  However this PR has been very helpful in finding a dozen errors in the cleanup. There was no need to merge it thought. But i thought it would actually help you: There where several complaints that new test cases did not use textblocks but less readable concatenations. And that did not stop because new testcases are typically copy&paste from old tests. Now some disagree. that's fine. lesson learned.

**Comment by stephan-herrmann** [comment]
> @stephan-herrmann may be you applied the cleanup manually with an old version? There was an error that was already fixed due to this PR: #2315

The failure occurred reproducibly by only pulling the change and running tests in the IDE. Seeing that it did not occur in other environments I will check if perhaps I was lacking a fix in the compiler, so not the source of the test but perhaps its compiled bytecode was wrong?

**Comment by jukzi** [comment]
> compiled bytecode was wrong?

yes, this particular test failed due to a error in compiler

**Comment by stephan-herrmann** [comment]
Before I propose changes to CONTRIBUTING.md, one question, @iloveeclipse  said:
> Every JDT committer should have received emails about new PR's created.

In fact I hadn't received any email about this PR. I quite liked the relative silence in my inbox and people know they can get my attention by mentioning me.  Apparently my github settings did not conform to general expectations, so I guess those expectations should be made explicit, too: What level of notifications / watching do we expect from committers?  I can select to watch "Pull requests" events, but it doesn't tell me if that's only for creation of new ones or all events touching any PR of the repo.

**Comment by jukzi** [comment]
Nobody is forced to get notifications other then the mailing list. You can either subscribe to github notifications on a repository to get notifications by mail or pull activities (new issues and Pull Requests) in the github ui on demand.
https://github.com/issues?q=is%3Aopen+user%3Aeclipse-jdt+
I for example have not subscribed but read other activities from time to time.

**Comment by srikanth-sankaran** [comment]
I have never been able to figure out on what basis I get notifications. These basically look random to me. In the past I have tried unsubscribing to specific threads - that never worked either. I doubt that I get notification on every PR. I seem to get on some.

**Comment by jukzi** [comment]
Have you tried to set "All activity"?
image

**Comment by iloveeclipse** [comment]
> I have never been able to figure out on what basis I get notifications.

Go to https://github.com/eclipse-jdt/eclipse.jdt.core

Check what is selected here:
image

If you would like to subscribe to a subset only, go there:
image

and it looks like one can't unsubscribe from a specific issue/PR.

**Comment by jukzi** [comment]
> and it looks like one can't unsubscribe from a specific issue/PR.

That works if you click "unsubscribe" in the specific PR:
image

**Comment by iloveeclipse** [comment]
Note: I **personally** expect that every committer should watch for new issues and PR's, since everyone who is committer usually committed some code to JDT that could have caused a bug or requires a fix.

As a committer (which is a role different to contributor), one can't assume the bugs will find "right" assignee automagically :-).

**Comment by jukzi** [comment]
Would be a lot overhead if everybody looks on every other activity. Its more efficient - and sufficient - that some take the burden to identify the cause and notify the related commiter by mentioning him.  Actually i think you contribute a lot by doing this @iloveeclipse

**Comment by akurtakov** [comment]
> > Would be a lot overhead if everybody looks on every other activity.
> 
> Well, below is my personal understanding about having a committer vs contributor role:
> 
>     * As an occasional contributor I'm only interested in my personal PR's or issues, so I don't subscribe to everything.
> 
>     * As a project committer I'm part of the team that is supposed to keep the project running well and so I should be at least informed about bugs and/or PR's created, so I should subscribe to bugs & PR's.

There is one more layer above - Project lead. And for these IMO it's expected that they follow up all activity in the project.

**Comment by iloveeclipse** [comment]
> There is one more layer above - Project lead. And for these IMO it's expected that they follow up all activity in the project.

Sure, that's why I'm receiving so much spam :-)

**Comment by srikanth-sankaran** [comment]
> > * As a project committer I'm part of the team that is supposed to keep the project running well and so I should be at least informed about bugs and/or PR's created, so I should subscribe to bugs & PR's.
> 
> I'd love to follow that advice but enabling notifications from issues and PRs already spams me with notifications about all comments dropped into _existing_ issues/prs, too.

This came up again recently in offline discussions. 

I think volunteer committers are a different breed than a committer. Subjecting them to an avalanche of notifications is a disservice to them and in turn to the project. Lately I have been guilty causing the breakdown of the flood gates - if you will pardon the mixed imagery -  that has resulted in too many notifications 😧 

Volunteers can be expected to pull data periodically (for commits integrated into master this happens anyway).

Plus If there is a clear delineation of their responsibility areas (e.g Stephan: Generics & Type inference, static analysis, JPMS, OpenJDK EG interfacing & Spec issues for new JEPs ...)  then tickets can be assigned and @mentions can be targetted to avoid overwhelming them.

In summary, my preferred model for a volunteer committer would that they would receive notifications only for issues in their areas and PRs where the author wishes to consult them by @mentioning them.
